### PR TITLE
Networking simplification

### PR DESCRIPTION
**Pull request by tomaka**
This big unreviewable PR refactors the networking part of the code, and more precisely the layers that coordinate all the TCP/WebSocket connections together. The code concerning individual connections has barely been touched.

The main file that has been modified is `collection.rs`. This file contains a data structure that represents a set of connections.

Before this PR, this data structure was "atomic". All the methods were taking `&self` as parameter, meaning that multiple methods could be called at the same time, and many of these methods were asynchronous and could be interrupted. If you needed to inject data into a connection, this had to be done through the data structure.
This design made the code extremely difficult to understand, because of all the corner cases to handle, most notably around the fact that futures can be cancelled by the user before their completion.

After this PR, this data structure has been split in two: one with the set of connections, and one `ConnectionTask` for each individual connection. The set of connections has a queue of messages destined to the `ConnectionTask`s, and the `ConnectionTask`s have a queue of messages destined to the set. This is exposed in the APIs of these objects, and it is the role of the user to do the messages passing. Before this PR, all the "locking/multithreading strategy" was handled internally by the collection, while after this PR it needs to be done by the user.

Thanks to this change in paradigm, the data structures are no longer atomic and are now simple mutable state machines. You have getters and you have methods that modify the state, and that's it. This considerably simplifies the implementation.

In the same vein, `peers.rs` and `service.rs`, which are data structures built on top of `collection.rs`, have been modified in the same way.

This new paradigm is in theory slightly less optimal than the one before. Before this PR, locking was fine-grained. If multiple threads wanted to access the set of connections at the same time, they could call a method at the same time, and if their changes didn't overlap they would actually run at the same time. After this PR, if multiple threads want to access the set, each thread needs to lock a mutex around the entire set.
However, the complexity of the previous implementation, notably around cancellable futures, has lead to a lot of overhead as well, where for example operations in progress needed to be buffered so that it can be resumed later in case the user interrupts the operation.
Overall I think that this change is more than worth it.

### CODE REVIEW

**Comment by tomaka** [comment]
This PR isn't completely finished, but the only thing that remains to update is the code of the full node.

**Comment by tomaka** [comment]
As part of this PR, I've removed the `node-info` feature of the full node.
It leads to too much code duplication, and it is unclear right now how to properly de-duplicate this code.

**Comment by tomaka** [comment]
There are high changes that this PR introduces some bugs, and I admit that I don't have the courage to chase bugs for days, and would prefer to merge this so that I do other networking-related changes on top of it.
However, the changes introduced by this PR reduce the chances of bugs in the long term by making the code so much more simple.

**Comment by tomaka** [comment]
@melekes Do you want to review that? You can completely say no (and actually I'd expect you to say no)

**Review by melekes — COMMENTED** [review]
man this PR is massive 🙈 I've reviewed some code, but not everything. Probably okay to merge. I can review the logic later.

Also would be interesting to compare how introducing global lock affected the performance. Will the tracing or some other instrument (similar to what `go tool pprof` provides)

File: bin/full-node/src/run/network_service.rs [hunk 1]
```diff
@@ -147,12 +161,48 @@ struct Inner {
...
+    >,
+
+    messages_from_connections_tx:
+        mpsc::Sender<(service::ConnectionId, service::ConnectionToCoordinator)>,
+
+    messages_from_connections_rx:
+        mpsc::Receiver<(service::ConnectionId, service::ConnectionToCoordinator)>,
+
+    /// List of all block requests that have been started but not finished yet.
+    blocks_requests: HashMap<
```
[the remark this answers is not in the document]
Comment (tomaka) [reply bin/full-node/src/run/network_service.rs:189]: The network state machine sends back an event when a request is finished, but the API of the network service is instead using asynchronous methods, so this tracking is necessary to convert between the two different kinds of APIs.

File: bin/full-node/src/run/network_service.rs [hunk 2]
```diff
@@ -183,235 +233,83 @@ impl NetworkService {
...
-                                    %peer_id,
-                                    %error,
-                                    "protocol-error"
-                                );
-                            }
-                        }
-                    };
+        // Initialize the inner network service.
+        let inner = {
+            let (messages_from_connections_tx, messages_from_connections_rx) = mpsc::channel(64);
```
[the remark this answers is not in the document]
Comment (tomaka) [reply bin/full-node/src/run/network_service.rs:267]: Arbitrary. If it's too low, there might be more context switches than necessary because the tasks running the connections have to wait for the task running the network to do some work. If it's too high, messages will wait for a long time in queue, which adds latency and memory usage.

File: bin/full-node/src/run/network_service.rs [hunk 3]
```diff
@@ -790,68 +602,620 @@ pub enum InitError {
...
+                        if let (1, protocol::BlocksRequestConfigStart::Hash(block_hash)) =
+                            (config.desired_count.get(), &config.start)
+                        {
+                            Some(block_hash)
+                        } else {
+                            None
+                        },
+                    );
+
+                    // TODO: is it a good idea to await here while the lock is held and freezing the entire networking background task?
```
[the remark this answers is not in the document]
Comment (tomaka) [reply bin/full-node/src/run/network_service.rs:808]: Yeah it's a rhetorical question. The answer is surely no, and this should be fixed at some point but not worth at the moment since the full node is secondary.

File: bin/light-base/src/network_service.rs [hunk 4]
```diff
@@ -1005,109 +807,638 @@ pub enum Event {
...
+            Some(Some(v)) => v,
+            _ => break,
+        };
+
+        lock.network
+            .inject_connection_message(connection_id, message);
+    }
+
+    // Process the events that the coordinator has generated.
+    'events_loop: loop {
```
[the remark this answers is not in the document]
Comment (tomaka) [reply bin/light-base/src/network_service.rs:844]: Same remark

File: bin/full-node/src/run/network_service.rs [hunk 5]
```diff
@@ -421,29 +319,14 @@ impl NetworkService {
                         futures_timer::Delay::new(next_discovery).await;
                         next_discovery = cmp::min(next_discovery * 2, Duration::from_secs(120));
 
-                        match inner
+                        let mut lock = inner.guarded.lock().await;
```
[the remark this answers is not in the document]
Comment (tomaka) [reply bin/full-node/src/run/network_service.rs:322]: Well, `guarded` isn't much better. `guarded` just means "protected by a mutex".
In the context of this module, there's only one mutex in the entire code, so to me it's not a problem to just call that `lock`.

**Comment by tomaka** [comment]
> Also would be interesting to compare how introducing global lock affected the performance. Will the tracing or some other instrument (similar to what go tool pprof provides) can give an insight into lock contention in Rust?

I don't think we can actually measure the lock contention at the moment, for what it's worth.
In principle there can be lock contention issues when you have GrandPa running, and parachains, and broadcasting transactions, and maybe other things I don't think of.
But right now we only ever access the lock to process events and to start block requests. Starting a block requests happens once every 6 seconds.
In other words, it's going to be completely negligible.

As for the light node, it's single threaded, so it's not measurable either.

File: bin/full-node/src/run/network_service.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (melekes) [reply bin/full-node/src/run/network_service.rs:322]: except one can find `Guarded` struct above, but there's no `Lock`

File: src/libp2p/collection.rs [hunk 6]
```diff
@@ -217,29 +124,121 @@ impl ConnectionId {
...
+    request_response_protocols: Arc<[ConfigRequestResponse]>,
 
     /// See [`Config::ping_protocol`].
-    ping_protocol: String,
+    ping_protocol: Arc<str>,
+}
+
+struct Connection<TConn> {
+    /// `true` if the connection is still in its handshake phase.
+    handshaking: bool,
```
[the remark this answers is not in the document]
Comment (tomaka) [reply src/libp2p/collection.rs:229]: You're probably right but it's unfortunately a bit complicated. I've added a TODO in the code instead.

**Comment by tomaka** [comment]
I think I've addressed everything :+1:

**Review by melekes — APPROVED** [review]
